pkgs/checks/lib/src/extensions/core.dart

```diff
@@ -0,0 +1,121 @@
+import 'package:checks/context.dart';
+
+extension TypeChecks on Check<Object?> {
+  /// Expects that the value is assignable to type [T].
+  ///
+  /// If the value is a [T], returns a [Check<T>] for further expectations.
+  Check<T> isA<T>() {
+    return context.nest<T>('is a $T', (actual) {
+      if (actual is! T) {
+        return Extracted.rejection(
+            actual: literal(actual), which: ['Is a ${actual.runtimeType}']);
+      }
+      return Extracted.value(actual);
+    }, atSameLevel: true);
+  }
+}
+
+extension HasField<T> on Check<T> {
+  /// Extracts a property of the value for further expectations.
+  ///
+  /// Sets up a clause that the value "has [name] that:" followed by any
+  /// expectations applied to the returned [Check].
+  Check<R> has<R>(R Function(T) extract, String name) {
+    return context.nest('has $name', (T value) {
+      try {
+        return Extracted.value(extract(value));
+      } catch (_) {
+        return Extracted.rejection(
+            actual: literal(value),
+            which: ['threw while trying to read property']);
+      }
+    });
+  }
+
+  /// Checks the expectations invoked in [condition] against this value.
+  ///
+  /// Use this method when it would otherwise not be possible to check multiple
+  /// properties of this value due to cascade notation already being used in a
+  /// way that would conflict.
+  ///
+  /// ```
+  /// checkThat(something)
+  ///   ..has((s) => s.foo, 'foo').equals(expectedFoo)
+  ///   ..has((s) => s.bar, 'bar').that((b) => b
+  ///     ..isLessThan(10)
+  ///     ..isGreaterThan(0));
+  /// ```
+  R that<R>(R Function(Check<T>) condition) => condition(this);
```

Comment:
Have you considered naming this method `which` instead?

It might be not quite right in English though, if so, I apologize :-)

My motivations are mostly aesthetic, seeing `checkThat() ... that()` seems somehow wrong to me (in Russian "that" in `checkThat()` "проверить что" and "that" later "который" have absolutely different meaning). And `that` looks close to `then`.

Reply:
Could get behind `which` too, as a another non-native English speaker.

I know that `that` is restrictive and `which` is descriptive in prose, so `that` is probably "more correct".

Reading it with a predicate
```dart
  checkThat(o).has((o) => o.foo).that(isTrue)
  checkThat(o).has((o) => o.foo).which(isTrue)
```
I can't say one reads strictly better than the other, but I kind-of *like* reading `which` more.
Maybe it's just the repetition of `that`.

Also, in
```dart
checkThat(Box(actual)).that((v) {
  v > 1;
  v < 4;
});
```
the `that` would work better as `where`, because it's not checking a single predicate. I think `which` is slightly better than `that`, but we are firmly in "personal preference" land with that.

And while `that` is close to `then`, `which` is equally close to `while` or `where`.

TL;DR: Slight preference for `which`.

Non-suggestion: `st`! `checkThat(o).st((v) ...)`. :wink: